### PR TITLE
tikv-ctl: add diff feature to check difference between two dbs

### PR DESCRIPTION
**Pull request by choleraehyq**
@zhangjinpeng1987 @BusyJay PTAL

### CODE REVIEW

File: src/bin/tikv-ctl.rs [hunk 1]
```diff
@@ -352,6 +368,80 @@ fn dump_raft_log_entry(db: DB, idx_key: &[u8]) {
...
+                 common_head_len,
+                 common_tail_len);
+
+        if !iter.valid() && iter2.valid() {
+            println!("db2: {:?}", escape(iter2.key()));
+            panic!("iter1 invalid but iter2 valid!");
+        }
+        if iter.valid() && !iter2.valid() {
+            println!("db1: {:?}", escape(iter.key()));
+            panic!("iter1 valid but iter2 invalid!");
```
Comment (zhangjinpeng1987) [inline src/bin/tikv-ctl.rs]: No panic, print all the remain keys.

File: src/bin/tikv-ctl.rs [hunk 2]
```diff
@@ -352,6 +375,86 @@ fn dump_raft_log_entry(db: DB, idx_key: &[u8]) {
...
+    let end_key = &keys::data_end_key(region.get_end_key());
+    for cf in ALL_CFS {
+        let handle = db.cf_handle(cf).unwrap();
+        let handle2 = db2.cf_handle(cf).unwrap();
+        println!("cf: {}", cf);
+        let mut ropt = ReadOptions::new();
+        ropt.set_iterate_upper_bound(end_key);
+        let mut iter = db.iter_cf_opt(handle, ropt);
+        let mut ropt = ReadOptions::new();
+        ropt.set_iterate_upper_bound(end_key);
```
Comment (siddontang) [inline src/bin/tikv-ctl.rs]: no need to set total_order_seek?

Comment (zhangjinpeng1987) [inline src/bin/tikv-ctl.rs]: no need

File: src/bin/tikv-ctl.rs [hunk 3]
```diff
@@ -352,6 +375,86 @@ fn dump_raft_log_entry(db: DB, idx_key: &[u8]) {
...
+                if iter.key() < iter2.key() {
+                    has_diff = true;
+                    println!("only db1 has : {:?}", escape(iter.key()));
+                    iter.next();
+                    continue;
+                }
+
+            }
+            if has_diff {
+                common_tail_len += 1;
```
Comment (AndreMouche) [inline src/bin/tikv-ctl.rs]: It seems `has_diff ` would always be true once different key appears?

File: src/bin/tikv-ctl.rs [hunk 4]
```diff
@@ -219,6 +237,11 @@ fn main() {
                 let _ = app.print_help();
             }
         }
+    } else if let Some(matches) = matches.subcommand_matches("diff") {
+        let region = String::from(matches.value_of("region").unwrap());
```
Comment (BusyJay) [inline src/bin/tikv-ctl.rs]: `matches.value_of().unwrap().parse().unwrap()`.

File: src/bin/tikv-ctl.rs [hunk 5]
```diff
@@ -352,6 +375,86 @@ fn dump_raft_log_entry(db: DB, idx_key: &[u8]) {
...
+            }
+            if has_diff {
+                common_tail_len += 1;
+            } else {
+                common_head_len += 1;
+            }
+            iter.next();
+            iter2.next();
+        }
+        println!("head have {} same keys, tail have {} same keys",
```
Comment (BusyJay) [inline src/bin/tikv-ctl.rs]: `common_tail_len` is meaningless.

File: src/bin/tikv-ctl.rs [hunk 6]
```diff
@@ -352,6 +375,84 @@ fn dump_raft_log_entry(db: DB, idx_key: &[u8]) {
...
+                    has_diff = true;
+                    println!("only db2 has : {:?}", escape(iter2.key()));
+                    println!("timestamp: {}",
+                             Key::from_encoded(iter2.key().to_vec()).decode_ts().unwrap());
+                    iter2.next();
+                    continue;
+                }
+                if iter.key() < iter2.key() {
+                    has_diff = true;
+                    println!("only db1 has : {:?}", escape(iter.key()));
```
Comment (siddontang) [inline src/bin/tikv-ctl.rs]: here you use db1 but above you use db, you should use db1 uniformly.

**Comment by siddontang** [comment]
PTAL @BusyJay

File: src/bin/tikv-ctl.rs (hunk 6 above)
Comment (choleraehyq) [inline src/bin/tikv-ctl.rs]: Done.

File: src/bin/tikv-ctl.rs [hunk 7]
```diff
@@ -199,6 +209,14 @@ fn main() {
         let start_ts = matches.value_of("start_ts").map(|s| s.parse().unwrap());
         let commit_ts = matches.value_of("commit_ts").map(|s| s.parse().unwrap());
         println!("You are searching Key {}: ", key);
+        let encoded_prefix = if key_encoded {
+            unescape(key)
+        } else {
+            encode_bytes(unescape(key).as_slice())
+        };
+        let data_key = keys::data_key(&encoded_prefix);
```
Comment (BusyJay) [inline src/bin/tikv-ctl.rs]: This is unnecessary.

File: src/bin/tikv-ctl.rs [hunk 8]
```diff
@@ -199,6 +209,14 @@ fn main() {
         let start_ts = matches.value_of("start_ts").map(|s| s.parse().unwrap());
         let commit_ts = matches.value_of("commit_ts").map(|s| s.parse().unwrap());
         println!("You are searching Key {}: ", key);
+        let encoded_prefix = if key_encoded {
+            unescape(key)
+        } else {
+            encode_bytes(unescape(key).as_slice())
```
Comment (BusyJay) [inline src/bin/tikv-ctl.rs]: You can only get ts from the encoded key. /cc @disksing

File: src/bin/tikv-ctl.rs [hunk 9]
```diff
@@ -352,6 +375,84 @@ fn dump_raft_log_entry(db: DB, idx_key: &[u8]) {
...
+        let mut iter2 = db2.iter_cf_opt(handle2, ropt);
+        iter.seek(SeekKey::Key(start_key));
+        iter2.seek(SeekKey::Key(start_key));
+        let mut has_diff = false;
+        let mut common_head_len = 0;
+        while iter.valid() && iter2.valid() {
+            if iter.key() != iter2.key() {
+                if iter.key() > iter2.key() {
+                    has_diff = true;
+                    println!("only db2 has : {:?}", escape(iter2.key()));
```
Comment (BusyJay) [inline src/bin/tikv-ctl.rs]: `:` is redundant.

File: src/bin/tikv-ctl.rs [hunk 10]
```diff
@@ -352,6 +367,84 @@ fn dump_raft_log_entry(db: DB, idx_key: &[u8]) {
...
+
+        if !iter.valid() && iter2.valid() {
+            println!("iter1 invalid but iter2 valid!");
+            while iter2.valid() {
+                println!("only db2 has : {:?}", escape(iter2.key()));
+                iter2.next();
+            }
+        }
+        if iter.valid() && !iter2.valid() {
+            println!("iter2 invalid but iter valid!");
```
Comment (siddontang) [inline src/bin/tikv-ctl.rs]: iter1 valid

File: src/bin/tikv-ctl.rs [hunk 11]
```diff
@@ -352,6 +367,84 @@ fn dump_raft_log_entry(db: DB, idx_key: &[u8]) {
...
+        iter.seek(SeekKey::Key(start_key));
+        iter2.seek(SeekKey::Key(start_key));
+        let mut has_diff = false;
+        let mut common_head_len = 0;
+        while iter.valid() && iter2.valid() {
+            if iter.key() != iter2.key() {
+                if iter.key() > iter2.key() {
+                    has_diff = true;
+                    println!("only db2 has : {}", escape(iter2.key()));
+                    println!("timestamp: {}",
```
Comment (siddontang) [inline src/bin/tikv-ctl.rs]: lock and raft have the timestamp?

**Comment by siddontang** [comment]
LGTM

**Comment by siddontang** [comment]
PTAL @zhangjinpeng1987

**Comment by zhangjinpeng1987** [comment]
LGTM

**Pull request merged**